prusti-contracts/src/ghost.rs

```diff
@@ -0,0 +1,147 @@
+use std::ops::*;
+use core::marker::PhantomData;
+
+pub unsafe auto trait Ghost {}
+
+macro_rules! implement_ghost_type {
+    ($ghost_type: ident) => {
+        pub struct $ghost_type;
+        unsafe impl Ghost for $ghost_type{}
+    };
+}
+
+macro_rules! implement_ghost_type_generic {
+    ($ghost_type: ident) => {
+        pub struct $ghost_type<T: Ghost> {
+            _type: PhantomData<T>
+        }
+        impl <T: Ghost> $ghost_type<T> {
+            ///  ```rust
+            ///  let seq_inst: GhostSeq<GhostInt> = GhostSeq::new();
+            ///  ```
+            pub fn new() -> Self {
+                $ghost_type {
+                    _type: PhantomData,
+                }
+            }
+        }
+        unsafe impl<T: Ghost> Ghost for $ghost_type<T> {}
+    }
+}
+
+// Ghost variant of Viper type, Int
+implement_ghost_type!(GhostInt);
+// wrappers around standard operations on GhostInt
+impl Add for GhostInt {
+    type Output = Self;
+    fn add(self, other: GhostInt) -> Self::Output {
+        GhostInt
+    }
+}
+
+impl Sub for GhostInt {
+    type Output = Self;
+    fn sub(self, other: GhostInt) -> Self::Output {
+        GhostInt
+    }
+}
+
+// Ghost variant of Viper type, Bool
+implement_ghost_type!(GhostBool);
+// wrappers around standard operations on GhostBool
+impl BitAnd for GhostBool {
+    type Output = Self;
+    fn bitand(self, rhs: Self) -> Self::Output {
+        GhostBool
+    }
+}
+
+impl BitOr for GhostBool {
+    type Output = Self;
+    fn bitor(self, rhs: Self) -> Self::Output {
+        GhostBool
+    }
+}
+
+impl Not for GhostBool {
+    type Output = Self;
+    fn not(self) -> Self::Output {
+        GhostBool
+    }
+}
+
+// Ghost variant of Viper type, Seq[T]
+implement_ghost_type_generic!(GhostSeq);
+// wrappers around standard operations on GhostSeq
+impl<T: Ghost> GhostSeq<T> {
+    /// ```rust
+    /// let seq: GhostSeq<i32> = GhostSeq::new();
+    /// seq.push(10);
+    /// ```
+    pub fn push(self, to_add: T) -> Self {
+        GhostSeq::new()
+    }
+
+    /// ```rust
+    /// let seq: GhostSeq<i32> = GhostSeq::new();
+    /// seq.push(10);
+    /// seq.pop(10);
+    /// ```
+    pub fn pop(self, to_remove: T) -> Self {
+        GhostSeq::new()
+    }
+
+    /// ```rust
+    /// let seq1: GhostSeq<i32> = GhostSeq::new();
+    /// seq1.append(seq2);
```

Comment:
`chain` instead of `append`? I suggest to enable [doctests](https://doc.rust-lang.org/rustdoc/documentation-tests.html) in the `Cargo.toml` to check those snippets.

Regarding all these documentiations consisting entirely of code snippets, it would be better to have at least 1 line of textual explanation describing what the method does, before the snippet that shows how to use the method.

Reply:
Thanks for the suggestion, I've improved the documentation.